Add Speed property

SirJohnK opened the pull request:
Added Speed property to AudioPlayer and sample application. Verified on Windows, Android and iOS.

jfversluis commented on src/Plugin.Maui.Audio/AudioPlayer.android.cs:
```diff
@@ -30,7 +30,23 @@ public double Balance
+			}
+			catch (Exception)
+			{
+				//Ensures that the speed limitations on different devices are handled!
```
Not sure I understand this comment, why are we hiding exceptions? Should we at least consider logging it somehow so the user has some way of knowing something might be wrong?

SirJohnK replied:
> Some questions, nothing shocking, I would love to see if we could keep that Android API level on 21 though

The PlaybackParams and the ability to set speed was added in API level 23. To support API level 21 we need to use the older SoundPool player. So to solve it, we should use the MediaPlayer for >= API 23 and SoundPool for < API 23. This will affect all properties and playback. But I can fix it, if API level 21 should be supported!

SirJohnK replied:
> Not sure I understand this comment, why are we hiding exceptions? Should we at least consider logging it somehow so the user has some way of knowing something might be wrong?

During my tests on different Android devices, it seems like different devices has different threshold values for what speed settings that is supported. Exceeding these values result in an Exception. Not sure if there is a way to query the device for the speed threshold value!? We could add a specific exception, like a SpeedThresholdValueExceededException and throw that. Then a application could alert the user based on that exception.

SirJohnK replied:
After looking more into SoundPool, it seems like it may not be a good solution for API level <23. It decompress sound into memory and has a 1 mb limitation. Great for short and multiple simultaneous sounds less than 6s. Perfect for a game engine like Orbit 😉 but useless for podcast app. Not sure there is another solution for API level 21!?

jfversluis commented on src/Plugin.Maui.Audio/AudioPlayer.macios.cs:
```diff
@@ -96,6 +102,7 @@ public void Stop()
     bool PreparePlayer()
     {
         player.FinishedPlaying += OnPlayerFinishedPlaying;
+		player.EnableRate = true;
```
Does it have any implications of having this on when someone is not using it? Should we turn it off if rate == 1?

SirJohnK replied:
According the Apple Dev Doc, this needs to be set before call to PrepareToPlay method. I can not find any information that it will have any implications if turned on. To turn it on, only when needed will probably require to recreate player, set property and call PrepareToPlay. One solution could be that we add a optional SpeedActive property to the CreatePlayer method and only attempt to set speed if active on all platforms.

jfversluis commented on src/Plugin.Maui.Audio/IAudioPlayer.cs:
```diff
@@ -30,6 +30,11 @@ public interface IAudioPlayer : IDisposable
 	double Balance { get; set; }
 
+	///<Summary>
+	/// Gets or sets the playback speed where 1 is normal speed.
```
Are there minimum and maximum values on specific platforms?

SirJohnK replied:
Yes, as I found on Android, it seems like it. And with AVAudioPlayer on iOS, on values 0.5 - 2.0 are supported. So the idea with the SpeedThresholdValueExceededException, maybe is not such a bad idea!?
[SpeedThresholdValueExceededException ](url)

jfversluis commented on src/Plugin.Maui.Audio/Plugin.Maui.Audio.csproj:
```diff
@@ -11,7 +11,7 @@
 		<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'">14.2</SupportedOSPlatformVersion>
 		<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'maccatalyst'">14.0</SupportedOSPlatformVersion>
-		<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">21.0</SupportedOSPlatformVersion>
+		<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">23.0</SupportedOSPlatformVersion>
```
I assume this was intended? Are you using APIs that are only available in API 23? If at all possible I'd like to keep this on 21 so that we support the same version as .NET MAUI itself.

SirJohnK replied:
For the plugin, please see my previous reply. For the sample application, it was an result on the change on the Plugin.

jfversluis reviewed (commented):
Some questions, nothing shocking, I would love to see if we could keep that Android API level on 21 though

SirJohnK commented:
Made an attempt to address all of these questions/issues in my feature branch: [https://github.com/SirJohnK/Plugin.Maui.Audio/tree/Speed_Property_Adjustment](url)

- Exception handling
- Speed out of range handling with SpeedOutOfRangeException
- Reverted to Android API level 21 and added SupportedOSPlatform("Android23.0") for Speed Property

jfversluis commented:
Hey @SirJohnK sorry for the delay on this! Guess we need to set the Speed property higher 😏

I'm off this week, I will have a look as soon as I can next week when I get back. In the meantime I do see a build error with the latest code?

> D:\a\Plugin.Maui.Audio\Plugin.Maui.Audio\src\Plugin.Maui.Audio\AudioPlayer.shared.cs(3,36): error CS0535: 'AudioPlayer' does not implement interface member 'IAudioPlayer.Speed'

SirJohnK commented:
Hey @jfversluis , no problem! :) I have fixed the build error. BUT I am looking for response on my latest comment on me making a new feature branch called "**Speed_Property_Adjustment**" in my repo. Where I think I have addressed all your comment on my initial PR. I did not want to merge it into this PR before response from you.

jfversluis commented:
@SirJohnK thanks for your patience! Yeah just merge all your changes in here then this will be the source of truth :) it's a bit hard to parse the changes with a separate branch in the mix as well. And that's the beauty of git, we'll still have all the history anyway!

SirJohnK commented:
> @SirJohnK thanks for your patience! Yeah just merge all your changes in here then this will be the source of truth :) it's a bit hard to parse the changes with a separate branch in the mix as well. And that's the beauty of git, we'll still have all the history anyway!

@jfversluis , done! :)

jfversluis commented:
OK! Finally getting around to this. Thank you so much for your unlimited patience here @SirJohnK 

I think this all looks good and verified it works, the only thing I'm wondering is if we should just limit the rate to 0.5 and 2.0 ourselves instead of throwing exceptions? @bijington any ideas around that?

On the other hand, it might be fine, I'll add it to the API docs regardless.

bijington commented:
Yes I would prefer to avoid throwing an exception in this situation if we can help it. As I understand iOS is potentially the most limiting right now with the range of 0.5 - 2.0? Is there a way we can programmatically access the max and min values for each platform and then use those to `Math.Clamp`?

jfversluis commented:
@SirJohnK let us know if this is something you'd still like to work on. If not, that is totally understandable after the time we took with this. Then I will see to take it home.

Thanks!

bijington commented:
Yes apologies for taking so long to look at this

jfversluis commented:
What do you think like this @bijington ?

Do we want to keep the exceptions for when `CanSetSpeed` is false?

bijington commented:
> What do you think like this @bijington ?
> 
> Do we want to keep the exceptions for when `CanSetSpeed` is false?

Yeah I am fine with that. If the developer can programmatically determine that the feature is disabled then that is ok. Just like how `FileStream.Seek` can `CanSeek` work.

bijington commented on src/Plugin.Maui.Audio/IAudioPlayer.cs:
```diff
@@ -30,6 +30,22 @@ public interface IAudioPlayer : IDisposable
+	///<Summary>
+	/// Gets a value indicating whether the playback speed can be changed.
+	///</Summary>
+	bool CanSetSpeed { get; }
```
Should we add `double MinimumSpeed` and `double MaximumSpeed` properties that each platform will return? Then the sample app could use those values for the max/min values of the `Slider`?

jfversluis replied:
Ooh I love that idea! Added!

bijington commented on src/Plugin.Maui.Audio/Plugin.Maui.Audio.csproj:
```diff
@@ -37,6 +37,9 @@
 	</PropertyGroup>
 
 	<!-- iOS & MacCatalyst -->
+	<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net6.0-ios|AnyCPU'">
+	  <CreatePackage>false</CreatePackage>
+	</PropertyGroup>
```
I don't think we want this? Pesky tooling

bijington reviewed (commented):
It's looking really close now! Just 2 items from me now

SirJohnK commented:
> @SirJohnK let us know if this is something you'd still like to work on. If not, that is totally understandable after the time we took with this. Then I will see to take it home.
> 
> Thanks!

I really like the changes! Much cleaner than my solution. Thanks for letting me play! ;)

bijington commented:
@jfversluis where are we at with this? Does it need another review?

jfversluis commented:
Yep! Please review and merge if you like what you see

bijington commented on src/Plugin.Maui.Audio/AudioPlayer.android.cs:
```diff
@@ -31,7 +32,37 @@ public double Balance
+                // Speed on Android can be between 0 and 6
+                var speedValue = Math.Clamp((float)value, 0.0f, 6.0f);
+
+                if (speedValue == float.NaN)
```
Match mac/iOS checks?

``suggestion
                if (float.IsNaN(speedValue))
```

bijington commented on src/Plugin.Maui.Audio/AudioPlayer.net.cs:
```diff
@@ -29,4 +29,12 @@ partial class AudioPlayer : IAudioPlayer
+
+	public double MinimumSpeed { get; }
+
+	public double MaximumSpeed { get; set; }
```
Do we need the setter?

bijington replied:
If not then here's a 'quick fix'

```suggestion
	public double MaximumSpeed { get; }
```

jfversluis replied:
Oh no definitely not, probably some muscle memory or something going on here, thanks :D

bijington reviewed (approved):
Just the point on setter for MaximumSpeed but the rest looks good!
